domain/unitstate/state/types.go

```diff
@@ -17,31 +17,63 @@ type unitName struct {
 }
 
 // unitState contains a YAML string representing the
-// state for a unit's uniter, storage or secrets.
+// state for a unit's uniter, storage and secrets.
 type unitState struct {
-	// State is the YAML string.
-	State string `db:"state"`
+	// UniterState is the units uniter state YAML string.
+	UniterState string `db:"uniter_state"`
+	// StorageState is the units storage state YAML string.
+	StorageState string `db:"storage_state"`
+	// SecretState is the units secret state YAML string.
+	SecretState string `db:"secret_state"`
 }
 
 // unitStateVal is a type for holding a key/value pair that is
 // a constituent in unit state for charm and relation.
-type unitStateKeyVal struct {
-	UUID string `db:"unit_uuid"`
-	// TODO (manadart 2024-09-09): This should be a generic T congruent with
-	// the function below. However, at the time of writing, SQLair does not
-	// support generic argumentation.
-	Key   any    `db:"key"`
+type unitStateKeyVal[T any] struct {
+	UUID  string `db:"unit_uuid"`
+	Key   T      `db:"key"`
 	Value string `db:"value"`
 }
 
-func makeUnitStateKeyVals[T comparable](unitUUID string, kv map[T]string) []unitStateKeyVal {
-	keyVals := make([]unitStateKeyVal, 0, len(kv))
+type unitCharmStateKeyVal unitStateKeyVal[string]
+type unitRelationStateKeyVal unitStateKeyVal[int]
```

Comment:
If Sqlair supported generic type arguments, these could be type _aliases_, and the methods below all generic.

Reply:
Type aliases don't work with SQLair because they are not known at runtime. As far as reflection is concerned its the original type, no record of the aliased name is stored.

I spent far too long trying to get this to work nicely, and even started to put it into SQLair, though there are a few trip ups. In the end I just settled with whats here, it barely benefits at all from being generic sadly.

https://github.com/canonical/sqlair/issues/162#issuecomment-2376608059